tutorials/haskell/hspec-golden-test-tutorial/tutorial.md

```diff
@@ -0,0 +1,506 @@
+---
+title: Golden Testing with Hspec-golden
+published: 2019-10-31
+ghc: 8.0.2
+lts: 14.10
+tags: haskell, golden, tests
+libraries: hspec-golden-0.1.0.1
+language: haskell
+author-name: Jorge Andrés Guerra Landázuri
+twitter-profile: Jorge_257
+github-profile: Jagl257
+description: Software testing is necessary to check if an application behaves in an expected way. Sometimes it is useful to store the expected output in a separate file. This is how golden tests work and in this tutorial we will be using the Hspec-golden haskell library to easily create and update our golden tests.
+---
+
+
+The value of software testing relies on the ability to check if an application behaves properly. Unit tests are the first layer of testing and help developers detect bugs early in the code by testing each component individually.
+
+For example, let's create a function that welcomes someone to this tutorial:
+
+```haskell
+sayHi :: String -> String
+sayHi name = "Welcome to the Golden Tests tutorial " ++ name
+```
+
+Using `hspec` we can easily test our function :
+
+```haskell
+import HelloWorld
+import Test.Hspec
+
+spec :: Spec
+spec = do
+    describe "sayHi" $ 
+        it "shows a Hello Golden Testers string" $
+            (sayHi "John Doe") `shouldBe` "Welcome to the Golden Tests tutorial John Doe"
+```
+
+In a nutshell, we are comparing the output of the function with a string that is stored inside the body of our test. We can take a different approach and store this expected output in a separate file. This approach is known as golden testing and the file in which we store the expected output takes the name of "golden file".
+
+So, now we know what golden testing is, but if unit tests and golden tests are so similar, why don't we just keep up with unit tests? Right?
+
+
+The truth is that unit tests are not useful when evaluating large, complex outputs and that's where the difference lies. Tests are all about maintainability and readability, and when we test large outputs we tend to add visual noise to our code. Golden Tests helpfully fix this issue. Let's check the main advantages:
+
+ - Large outputs are stored in a golden file reducing the visual noise in our code.
+
+ - With some golden tests libraries the Golden Files can be automatically generated.
+
+ - Some golden test libraries also automatically update the golden file. So we don't need to manually update the tests when the expected output changes. 
+
+ - It is useful when working with JSON, HTML, images, etc.
+
+In this tutorial we will use [hspec-golden](http://hackage.haskell.org/package/hspec-golden) library to build golden tests. We will also use the `hspec-golden` CLI to easily update them.
```

Comment:
Let's move all the links to the bottom of the file using reference links instead of inline links:

```suggestion
[hspec-golden][hspec-golden-link]

[hspec-golden-link]: http://hackage.haskell.org/package/hspec-golden
```